### ISSUE TITLE
improve webpack3 build performance

### ISSUE THREAD

**Issue by williamstein** [sagemathinc/cocalc#2629]
See https://webpack.js.org/guides/build-performance/

What we care about is:

1. Start webpack watch going.
2. Change one character in a little cjsx file and save it
3. Wait about 15s for the build to finish...

We want that 15s to be about 3s, like it has been in the past.

**Comment by williamstein** [comment]
Bottom of this issue suggests a fix that sped things up by 100x for them:

https://github.com/webpack/webpack/issues/4558

**Comment by johnjeng** [comment]
Unfortunately that one concerns uglifyjs which we don't use on development. Though it will likely speed up production builds.

**Comment by williamstein** [comment]
My guess after looking into this more is that the best approach for us is 

https://webpack.js.org/plugins/dll-plugin/#dllreferenceplugin

This would take significant work to implement.  There's a nice blog post about this approach at https://robertknight.github.io/posts/webpack-dll-plugins/, but it's about webpack1.

**Comment by williamstein** [comment]
If you work on this it is critical to use the --profile option so you know what the heck is going on, e.g.,

```
$ SOURCE_MAP=true NODE_ENV=development webpack --debug --output-pathinfo --progress --colors --watch --profile
...

Make changes
...
See this:

170ms building modules
78ms sealing
0ms optimizing
0ms basic module optimization
19ms module optimization
25ms advanced module optimization
0ms basic chunk optimization
0ms chunk optimization
0ms advanced chunk optimization
0ms module and chunk tree optimization
0ms chunk modules optimization
1ms advanced chunk modules optimization
17ms module reviving
2ms module order optimization
10ms module id optimization
10ms chunk reviving
0ms chunk order optimization
13ms chunk id optimization
142ms hashing
2ms module assets processing
0ms chunk assets processing
5ms additional chunk assets processing
0ms recording
0ms additional asset processing
0ms chunk asset optimization
10184ms asset optimization
776ms emitting
Hash: 2802b2f40ebd5adfe876
Version: webpack 3.10.0
Time: 11841ms
```

**Comment by williamstein** [comment]
This thread is about people experiencing massive slowdowns when using webpack3 instead of webpack2, with no solution in site: https://github.com/webpack/webpack/issues/5277

**Comment by williamstein** [comment]
I tried https://github.com/mzgoddard/hard-source-webpack-plugin with the cache on a ramdisk even, by adding this code:
```
if DEVMODE
    cache = '/tmp/webpack-cache/[confighash]'
    console.log 'using hard-source plugin for better caching -- cache at ', cache
    HardSourceWebpackPlugin = require('hard-source-webpack-plugin')
    plugins.push(new HardSourceWebpackPlugin(cacheDirectory:cache))
```
The cache is definitely used for something.  But it makes no difference, since almost all time is spent in "asset optimization", and this is no faster.

**Comment by williamstein** [comment]
OK, I got to the bottom of this!  There is an array of plugins called `staticPages`.  Each one is very complicated, and I think converts .pug templates to html, and has lots of options.  It's maybe a hack where this is a separate plugin for every file.   All I did was remove the one line where (all these) staticPages plugins are included, and the re-build time dropped from about 15-20s down to 2s.  See:
```
9ms chunk reviving
0ms chunk order optimization
11ms chunk id optimization
150ms hashing
5ms module assets processing
199ms chunk assets processing
6ms additional chunk assets processing
0ms recording
0ms additional asset processing
0ms chunk asset optimization
1086ms asset optimization
206ms emitting
Hash: 52bafa388c4299c9d4d0
Version: webpack 3.10.0
Time: 1995ms
```

I'm pushing this being disabled to master.  It **breaks** doing dev if you don't visit .../app directly, so we must all be aware of this until this is properly fixed.
